### ISSUE TITLE
--fp16 utlizes significantly higher memory and results in OOM

### ISSUE THREAD

**Issue by medabalimi** [pytorch/fairseq#2907]
#### What is your question?
fp16 seems to take up  >2x memory on GPU compared to fp32. I have been under the assumption that fp16 in addition to be faster is more memory optimized as well. using --memory-efficient-fp16 does reduce memory optimization (slightly) to less than fp32. I'm using the wav2vec 2.0 examples

#### What have you tried?
Pytorch 1.6.0, Pytorch 1.7.0 prebuilt wheels with support for CUDA 10.2
Pytorch 1.7.0 prebuilt wheel with support for CUDA 11.0
Pytorch 1.7.0 from source with support for CUDA Compute 3.5

#### What's your environment?

 - fairseq Version: commit 108f7204f6ccddb676e6d52006da219ce96a02dc
 - PyTorch Version: 1.7.0
 - OS: Ubuntu 20.04
 - How you installed fairseq : source
 - Build command you used (if compiling from source): pip install -e .
 - Python version: 3.8.5
 - CUDA/cuDNN version: 11.0/8.0.5.39-1 (same issue with CUDA 10.2 and cuDNN 7.6.5)
 - GPU models and configuration: RTX 2070 Super, Tesla K40 (pytorch 1.7.0 was compiled with support for sm3.5)
 - Any other relevant information:
https://github.com/huggingface/transformers/issues/8403 -- seems to indicate issues exist with pytorch versions 1.6 and later (likely due to native AMP)
https://discuss.pytorch.org/t/fp16-training-with-feedforward-network-slower-time-and-no-memory-reduction/95560

I notice that memory usage grows whenever overflow is detected and loss scale is changed. This could just be a coincidence though.

**Comment by thpun** [comment]
Just fyi: [Nvidia doc](https://docs.nvidia.com/deeplearning/performance/mixed-precision-training/index.html#prereqs) was written that mixed precision training requires Volta or Turing architecture. So i dont think K40 supports FP16, but RTX2070 Super should support that.

**Comment by medabalimi** [comment]
I'm aware that K40 does not support FP16, but AFAIK one can still do fp16 on non supported architectures/devices with no real performance benefit (it may even be slower like 10 series GPUs).  The issue I'm raising has to  do with increased memory utilization with --fp16 enabled.

**Comment by myleott** [comment]
`--fp16` enables mixed precision training, where we maintain both an FP16 and FP32 copy of the model weights (the latter is used in the optimizer), and all activations/gradients are stored in FP16.

Thus you will see increased memory usage from model weights (1.5x larger than FP32 alone), but less memory usage from the forward/backward pass activations/gradients.

`--memory-efficient-fp16` gets rid of the FP32 model copy and only maintains FP32 momentum in the optimizer. Thus you'll see 0.5x memory usage from the model weights, 0.5x memory usage in the forward/backward, and 1.0x memory usage in the optimizer (relative to FP32).

**Comment by Oxi84** [comment]
> 
> 
> `--fp16` enables mixed precision training, where we maintain both an FP16 and FP32 copy of the model weights (the latter is used in the optimizer), and all activations/gradients are stored in FP16.
> 
> Thus you will see increased memory usage from model weights (1.5x larger than FP32 alone), but less memory usage from the forward/backward pass activations/gradients.
> 
> `--memory-efficient-fp16` gets rid of the FP32 model copy and only maintains FP32 momentum in the optimizer. Thus you'll see 0.5x memory usage from the model weights, 0.5x memory usage in the forward/backward, and 1.0x memory usage in the optimizer (relative to FP32).

This sounds amazing for --memory-efficient-fp16 argument, but how would i do that if i do not not load from command line but using these lines

    from fairseq.models.transformer import TransformerModel
    de2en = TransformerModel.from_pretrained(.....)
    de2en.half()

I use half to make it fp16 do not know how to make it --memory-efficient-fp16
